lib/curl_libproxy.c

```diff
@@ -0,0 +1,75 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2011 - 2016, Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at http://curl.haxx.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#ifdef ENABLE_LIBPROXY
+
+#include "urldata.h"
+
+#include <proxy.h>
+/* The last 3 #include files should be in this order */
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+
+static pxProxyFactory *factory;
+
+CURLcode Curl_libproxy_global_init(void)
+{
+  factory = px_proxy_factory_new();
+  if(!factory)
+    return CURLE_OUT_OF_MEMORY;
+
+  return CURLE_OK;
+}
+
+void Curl_libproxy_global_cleanup(void)
+{
+  if(factory)
+    px_proxy_factory_free(factory);
+
+  factory = NULL;
+}
+
+char *Curl_libproxy_detect_proxy(const char *url)
+{
+  char *result = NULL;
+
+  if(factory) {
+    char **libproxy_results = px_proxy_factory_get_proxies(factory, url);
+
+    if(libproxy_results) {
+      int i;
+
+      /* We only cope with one; can't fall back on failure */
+      result = libproxy_results[0];
+      for(i=1; libproxy_results[i]; i++)
+        free(libproxy_results[i]);
+      free(libproxy_results);
+    }
+  }
+
+  return result;
```

Comment:
1. When is this pointer free'd again?
2. libproxy doesn't return the proxy URL using the same schemes as curl supports does it? I mean for SOCKS4/4a/5/5h does it?
3. Can a libproxy proxy be used over unix domain sockets?

Reply:
I'll let jbrummer answer 1.

Regarding 2, libproxy may return `socks://` (which means: try `socks5://`, then `socks4a://`, then `socks4://`, in that order), or it may return `direct://` (which means "connect directly"). These need to be translated into curl proxy syntax. Then I don't know about `socks5h://` because GLib doesn't know about `socks5h://`... but I guess that should that be attempted first, before `socks5://`?

Regarding 3. libproxy just returns URIs. It would not generally make sense to connect to an HTTP proxy via anything other than TCP, but it's theoretically possible, if you are running a proxy server locally and have taught it to respond via UNIX domain sockets rather than TCP. Same for the other common proxy types, HTTPS and FTP and SOCKS (at least I think SOCKS goes over TCP?). But in theory, if you pass a different URI scheme, it could return any sort of proxy URI.

Reply:
> libproxy may return socks:// (which means: try socks5://, then socks4a://, then socks4://, in that order)

BTW this does not mean libproxy might not return more specific URI schemes too. libproxy has basically no documentation, so better assume it will return whatever it wants. ;) But specifically `socks://` and `direct://` are going to require special handling here. In particular, if you do not see `direct://` then you must not connect directly.